File: cloudflare/alife-app/src/utils/eventPackagePolicyEditor.ts
import type { EventPackagePolicyAdmin, PolicyEditorDefaults, PolicyRules, PolicyTier, PublishEventPackagePolicyRequest } from '../types/eventPackagePolicyAdmin'

// Presentation order follows if / else if / else; stored rule order is preserved.
export const policyTiers: PolicyTier[] = ['enhanced', 'standard', 'light']
export const noCurrentPolicy = '00000000-0000-0000-0000-000000000000'
export const isCurrentPolicy = (p: EventPackagePolicyAdmin, now = Date.now()) => p.isPublished && Date.parse(p.effectiveFromUtc) <= now && (!p.retiredUtc || Date.parse(p.retiredUtc) > now)
export const validityDays = (value: string): number => {
  const match = /^P(?:(\d+(?:\.\d+)?)D)?(?:T(?:(\d+(?:\.\d+)?)H)?(?:(\d+(?:\.\d+)?)M)?(?:(\d+(?:\.\d+)?)S)?)?$/.exec(value)
  if (!match) throw new Error('Unsupported approval duration')
  const days = Number(match[1] || 0) + Number(match[2] || 0) / 24 + Number(match[3] || 0) / 1440 + Number(match[4] || 0) / 86400
  if (!(days > 0)) throw new Error('Invalid approval duration')
  return days
}

// Unknown structures are read-only: never round-trip a partially understood policy.
export const readPolicyRules = (input: Record<string, unknown>, catalog: PolicyEditorDefaults): PolicyRules => {
  const r = structuredClone(input) as PolicyRules
  const shape = (value: object, keys: string[], optional: string[] = []) => {
    if (!value || Array.isArray(value) || typeof value !== 'object' || keys.some(k => !(k in value)) || Object.keys(value).some(k => !keys.includes(k) && !optional.includes(k))) throw new Error('Unsupported policy')
  }
  shape(r, Object.keys(catalog.rules))
  if (r.schemaVersion !== '1' || !Number.isInteger(r.preEventConfirmationWindowHours) || r.preEventConfirmationWindowHours <= 0 || typeof r.conditionWaiverAllowed !== 'boolean' || !Array.isArray(r.materialChangeRules)) throw new Error('Unsupported policy')
  if (!Array.isArray(r.tierRules) || r.tierRules.length !== 3 || policyTiers.some(t => r.tierRules.filter(x => x.tier === t).length !== 1)) throw new Error('Unsupported tiers')
  shape(r.authorityByTier, policyTiers); shape(r.approvalValidityByTier, policyTiers)
  for (const tier of r.tierRules) {
    shape(tier, ['tier', 'whenAnyConfirmedFactCodes', 'whenAnyActivityTypeCodes', 'whenAnyModuleCodes'])
    for (const [key, options] of [['whenAnyConfirmedFactCodes', catalog.facts], ['whenAnyActivityTypeCodes', catalog.activityTypes], ['whenAnyModuleCodes', catalog.modules]] as const) {
      if (!Array.isArray(tier[key]) || tier[key].some(code => !options.some(o => o.code === code)) || new Set(tier[key]).size !== tier[key].length) throw new Error('Unavailable trigger')
    }
    shape(r.authorityByTier[tier.tier], ['minimumApproverCount'])
    const count = r.authorityByTier[tier.tier].minimumApproverCount
    if (!Number.isInteger(count) || count < 1 || count > 5) throw new Error('Invalid approver count')
    validityDays(r.approvalValidityByTier[tier.tier])
  }
  shape(r.delegationRules, ['enabled'], ['allowedTiers'])
  if (typeof r.delegationRules.enabled !== 'boolean' || (r.delegationRules.allowedTiers ?? []).some(t => !policyTiers.includes(t))) throw new Error('Unsupported delegation')
  shape(r.legacyRollout, ['effectiveFromUtc', 'transitionDeadlineUtc', 'cohortRule', 'safetyCriticalModuleCodes', 'transitionByMode'])
  if (r.legacyRollout.cohortRule !== 'new-events-first' || !Array.isArray(r.legacyRollout.safetyCriticalModuleCodes) || r.legacyRollout.safetyCriticalModuleCodes.some(c => !catalog.modules.some(m => m.code === c)) || !(Date.parse(r.legacyRollout.transitionDeadlineUtc) > Date.parse(r.legacyRollout.effectiveFromUtc))) throw new Error('Unsupported rollout')
  shape(r.legacyRollout.transitionByMode, [], ['off', 'dryRun', 'enforced'])
  for (const [mode, value] of Object.entries(r.legacyRollout.transitionByMode)) if (catalog.rules.legacyRollout.transitionByMode[mode] !== value) throw new Error('Unsupported transition')
  return r
}

export const createPolicyDraft = (catalog: PolicyEditorDefaults, source: EventPackagePolicyAdmin | undefined, current: EventPackagePolicyAdmin | undefined, now = new Date()): PublishEventPackagePolicyRequest => {
  const rules = readPolicyRules(source?.rules ?? catalog.rules, catalog)
  rules.legacyRollout.effectiveFromUtc = now.toISOString()
  rules.legacyRollout.transitionDeadlineUtc = new Date(now.getTime() + 90 * 86400000).toISOString()
  return { version: `v${now.toISOString().replace(/\D/g, '')}`, schemaVersion: '1', rules,
    enforcementMode: source?.enforcementMode ?? 'dryRun', effectiveFromUtc: now.toISOString(),
    expectedCurrentPolicyId: current?.id ?? noCurrentPolicy, sourcePolicyId: source?.id }
}

export const changedPolicySections = (before: Record<string, unknown> | undefined, after: PolicyRules): string[] => Object.keys(after).filter(k => k !== 'schemaVersion' && JSON.stringify(before?.[k]) !== JSON.stringify(after[k]))

File: backend/src/Alife.Application/Admin/EventPackagePolicies/EventPackagePolicyEditor.cs
using System.Text.Json;
using Alife.Application.Common.Interfaces;
using Alife.Application.Common.Models;
using Alife.Application.Events.Dtos;
using Alife.Application.Events.Services;
using Alife.Domain.Entities;
using Alife.Domain.Enums;
using MediatR;
using Microsoft.EntityFrameworkCore;

namespace Alife.Application.Admin.EventPackagePolicies;

public sealed record PolicyOption(string Code, LocalizedTextDto Label);
public sealed record PolicyEditorDefaults(JsonElement Rules, IReadOnlyList<PolicyOption> Facts,
    IReadOnlyList<PolicyOption> ActivityTypes, IReadOnlyList<PolicyOption> Modules);
public sealed record PolicyImpact(Guid? CurrentPolicyId, int AffectedEventCount, int AffectedApprovalCount, string ImpactToken);
public sealed record GetPolicyEditorDefaultsQuery(Guid MemberId) : IRequest<AppResult<PolicyEditorDefaults>>;
public sealed record PreviewPolicyQuery(Guid MemberId, PublishEventPackagePolicyRequest Request) : IRequest<AppResult<PolicyImpact>>;

public static class EventPackagePolicyEditor
{
    public static readonly PolicyOption[] Facts =
    [
        new("people.volunteersRequired", new("Volunteers required", "需要同工")),
        new("money.hasMoneyFlow", new("Money involved", "涉及费用收支")),
        new("safety.requiresRam", new("Risk assessment required", "需要风险评估")),
        new("people.childrenPresent", new("Children participating", "儿童参与")),
        new("programme.productionRequired", new("Managed programme", "需要节目制作")),
        new("place.resourcesRequired", new("Venue resources required", "需要场地资源")),
        new("move.transportRequired", new("Transport required", "需要交通安排")),
        new("move.accommodationRequired", new("Accommodation required", "需要住宿安排")),
        new("food.serviceRequired", new("Food service required", "需要餐饮服务")),
        new("scale.multiZone", new("Multiple operating zones", "多区域现场运营")),
        new("comms.followupRequired", new("Follow-up required", "需要跟进")),
        new("event.exists", new("Every event", "所有活动"))
    ];

    public static JsonElement DefaultRules(DateTime now) => JsonSerializer.SerializeToElement(new
    {
        schemaVersion = "1", preEventConfirmationWindowHours = 72,
        tierRules = new[] {
            Tier("light", [], [], []),
            Tier("standard", ["money.hasMoneyFlow", "place.resourcesRequired"], [], ["PEOPLE.REGISTRATION", "PLACE.RESOURCE", "MONEY.FINANCE"]),
            Tier("enhanced", ["people.childrenPresent", "move.transportRequired", "move.accommodationRequired", "safety.requiresRam"], ["outdoor-activity"], ["SAFEGUARDING.CHILD", "SAFETY.RAM", "MOVE.STAY", "FESTIVAL.OPERATIONS"])
        },
        authorityByTier = new { light = new { minimumApproverCount = 1 }, standard = new { minimumApproverCount = 1 }, enhanced = new { minimumApproverCount = 1 } },
        approvalValidityByTier = new { light = "P30D", standard = "P14D", enhanced = "P7D" },
        materialChangeRules = Array.Empty<string>(), conditionWaiverAllowed = false,
        delegationRules = new { enabled = false, allowedTiers = Array.Empty<string>() },
        legacyRollout = new { effectiveFromUtc = now, transitionDeadlineUtc = now.AddDays(90),
            cohortRule = "new-events-first", safetyCriticalModuleCodes = new[] { "SAFETY.RAM", "SAFEGUARDING.CHILD" },
            transitionByMode = new { off = "legacyReadOnlyPackage", dryRun = "timeLimitedCompatibility", enforced = "formalPackageRequired" } }
    });

    private static object Tier(string tier, string[] facts, string[] types, string[] modules) => new
    { tier, whenAnyConfirmedFactCodes = facts, whenAnyActivityTypeCodes = types, whenAnyModuleCodes = modules };

    internal static IQueryable<GroupEvent> AffectedEvents(IAlifeDbContext db, Guid? scope, DateTime now)
        => db.GroupEvents.Where(x => !x.IsDeleted && (scope.HasValue ? x.GroupId == scope.Value :
            !db.EventPackageGovernancePolicyVersions.Any(p => p.OrganisationId == x.GroupId && p.IsPublished &&
                p.EffectiveFromUtc <= now && (!p.RetiredUtc.HasValue || p.RetiredUtc > now))));

    internal static async Task<PolicyImpact> Impact(IAlifeDbContext db, Guid? scope, CancellationToken ct)
    {
        var now = DateTime.UtcNow;
        var current = await db.EventPackageGovernancePolicyVersions.AsNoTracking().Where(x => x.OrganisationId == scope &&
            x.IsPublished && x.EffectiveFromUtc <= now && (!x.RetiredUtc.HasValue || x.RetiredUtc > now))
            .OrderByDescending(x => x.EffectiveFromUtc).Select(x => (Guid?)x.Id).FirstOrDefaultAsync(ct);
        var events = await AffectedEvents(db, scope, now).Select(x => x.Id).OrderBy(x => x).ToListAsync(ct);
        var approvals = await db.EventPackages.AsNoTracking().Where(x => events.Contains(x.EventId) &&
            x.ApprovalValidityStatus == EventPackageApprovalValidity.Active).Select(x => x.Id).OrderBy(x => x).ToListAsync(ct);
        return new(current, events.Count, approvals.Count, EventPackageCanonicalizer.HashCanonical(new { current, events, approvals }));
    }
}

public sealed class GetPolicyEditorDefaultsHandler(IAlifeDbContext db, IEventActivityTemplateCatalog templates)
    : IRequestHandler<GetPolicyEditorDefaultsQuery, AppResult<PolicyEditorDefaults>>
{
    public async Task<AppResult<PolicyEditorDefaults>> Handle(GetPolicyEditorDefaultsQuery request, CancellationToken ct)
    {
        if (!await ListEventPackagePoliciesQueryHandler.CanManage(db, request.MemberId, ct))
            return AppResult<PolicyEditorDefaults>.Forbidden("Event Package policy administration permission is required.");
        var types = (await templates.ListAsync(true, ct)).ToDictionary(x => x.Definition.Code, x => x.Definition);
        return AppResult<PolicyEditorDefaults>.Success(new(EventPackagePolicyEditor.DefaultRules(DateTime.UtcNow),
            EventPackagePolicyEditor.Facts, types.Values.Select(x => new PolicyOption(x.Code, x.Name)).ToArray(),
            EventCompositionDefinitions.Modules.Select(x => new PolicyOption(x.Code, x.Name)).ToArray()));
    }
}

public sealed class PreviewPolicyHandler(IAlifeDbContext db, IEventActivityTemplateCatalog templates)
    : IRequestHandler<PreviewPolicyQuery, AppResult<PolicyImpact>>
{
    public async Task<AppResult<PolicyImpact>> Handle(PreviewPolicyQuery query, CancellationToken ct)
    {
        if (!await ListEventPackagePoliciesQueryHandler.CanManage(db, query.MemberId, ct))
            return AppResult<PolicyImpact>.Forbidden("Event Package policy administration permission is required.");
        var error = EventPackagePolicyRules.Validate(query.Request, (await templates.ListAsync(true, ct)).Select(x => x.Definition.Code).ToHashSet());
        if (error is not null) return AppResult<PolicyImpact>.Validation(error);
        var impact = await EventPackagePolicyEditor.Impact(db, query.Request.OrganisationId, ct);
        if (query.Request.ExpectedCurrentPolicyId.HasValue && query.Request.ExpectedCurrentPolicyId != (impact.CurrentPolicyId ?? Guid.Empty))
            return AppResult<PolicyImpact>.Conflict("The current policy changed. Refresh and review again.");
        return AppResult<PolicyImpact>.Success(impact);
    }
}
